### ISSUE TITLE
TensorBoard logging batch level metrics

### ISSUE THREAD

**Issue by sameermanek** [fchollet/keras#6692]
It'd be useful if there was some batch-level logging in TensorBoard when using the TensorBoard callback (as defined in keras/callbacks.py). I think this'd be generally useful when trying to keep track of stats between epochs. 

As an example, there could be a new boolean argument `write_batch_performance` in the __init__() method, and a new `on_batch_end` method, something like:

```
    def on_batch_end(self, batch, logs=None):
        logs = logs or {}

        if self.write_batch_performance == True:
            for name, value in logs.items():
                if name in ['batch','size']:
                    continue
                summary = tf.Summary()
                summary_value = summary.value.add()
                summary_value.simple_value = value.item()
                summary_value.tag = name
                self.writer.add_summary(summary, self.seen)
            self.writer.flush()

        self.seen += self.batch_size
```

I have a basic version of this locally; I'd need to slightly clean it up and incorporate it into the unit tests. Happy to do so if it makes sense to incorporate into keras. I couldn't find any matching outstanding feature requests. 

Thanks!

**Comment by sxs4337** [comment]
+1
This would be very useful indeed!
Specially when training on very large datasets.

**Comment by Barfknecht** [comment]
I have been looking for something like this for a week. @sameermanek , does your implementation perform well?

**Comment by sameermanek** [comment]
I haven't explicitly tested (computational) performance, but it has been useful to me (e..g, when testing things locally on CPU). 
The modifications I made locally are [here](https://github.com/sameermanek/keras/commit/7ceefaacd664f5065a06d76e80e7fccadf737db5). 
In terms of output, this'll log the batch-level performance, so I can see whether anything's going haywire relatively early on (clearly though, I should've stopped this one a little earlier than I did)
![example](https://user-images.githubusercontent.com/6383638/27703365-1eae2944-5cd5-11e7-9302-c8079b64a920.png)

**Comment by Barfknecht** [comment]
Thanks! This looks really great, and I agree, so much easier to see what is happening between epochs. I am currently having problems with a large dataset myself. So this is just a modification of the tensorboard callback? I will then use the callback as normal. Should I see live graphing of the batches or will that still only show every epoch.

**Comment by sameermanek** [comment]
@Barfknecht Correct -- this is just a modification of the tensorboard callback. I probably gave you the wrong link (sorry about that -- that was just a commit, not the entire file). The file is [here](https://github.com/sameermanek/keras/blob/7ceefaacd664f5065a06d76e80e7fccadf737db5/keras/callbacks.py) and you can see the added `write_batch_performance` argument for the Tensorboard callback (line 635).

You should see live graphing of the batches (example screenshot below; one epoch is 5,000 steps in this, so you can see there are values well before we get to an epoch)
![example](https://user-images.githubusercontent.com/6383638/27707213-7d7fb5c0-5ce2-11e7-88fd-6043ad3f9df0.png)

**Comment by wkretzsch** [comment]
This would be really useful!  Any ETA on when this will be merged into master? I don't see a PR?

**Comment by BrikerMan** [comment]
This is what I needed, thanks a lot.

**Comment by sameermanek** [comment]
@wkretzsch I just submitted a [PR](https://github.com/fchollet/keras/pull/7617); let's see whether it's accepted or if there's any feedback. Thanks.

**Comment by JulesGM** [comment]
bump. I'd really like to see this make it to master.

**Comment by PeterPanUnderhill** [comment]
Short notice for those who got only one dot in the graph: remember to set **write_batch_performance** to True!

**Comment by jpcenteno80** [comment]
Hopefully the PR will be approved soon... Since I didn't want to modify the keras/callbacks.py file, I tried to implement this by sub-classing from the `class TensorBoard(Callback)`:
```
import tensorflow as tf

class TensorBoard_and_write_batch_performance(TensorBoard):
    '''
    Writes batch performance to TensorBoard
    '''
    def on_batch_end(self, batch, logs=None):
        self.seen = 0

        for name, value in logs.items():
            if name in ['batch', 'size']:
                continue
            summary = tf.Summary()
            summary_value = summary.value.add()
            summary_value.simple_value = value.item()
            summary_value.tag = name
            self.writer.add_summary(summary, self.seen)
        self.writer.flush()
        self.seen += self.batch_size
        
        super(TensorBoard_and_write_batch_performance, self).on_batch_end(batch, logs)
```
However, when I run this, I can only view the batch performance if I use 'RELATIVE':
![image](https://user-images.githubusercontent.com/22669584/37308418-9cb64d60-2603-11e8-8dd1-75b2f1ed73c1.png)
And I also loose the 'STEP' tracker, so I don't know in which epoch I am.
Finally, I don't have access to the global variable `tf` and that is why I need to import it in my module.

Does anyone have any tips on how to bring back epoch tracker with this sub-class-TensorBoard version of write_batch_performance?

**Comment by VladislavZavadskyy** [comment]
@jpcenteno80, try this:
```python
class TB(callbacks.TensorBoard):
    def __init__(self, log_every=1, **kwargs):
        super().__init__(**kwargs)
        self.log_every = log_every
        self.counter = 0
    
    def on_batch_end(self, batch, logs=None):
        self.counter+=1
        if self.counter%self.log_every==0:
            for name, value in logs.items():
                if name in ['batch', 'size']:
                    continue
                summary = tf.Summary()
                summary_value = summary.value.add()
                summary_value.simple_value = value.item()
                summary_value.tag = name
                self.writer.add_summary(summary, self.counter)
            self.writer.flush()
        
        super().on_batch_end(batch, logs)
```
Your `seen` counter is reset every batch, so step is always zero.

**Comment by RyanCodes44** [comment]
Where exactly does this change have to occur? Tried changing it in:

`C:\Users\rc\AppData\Local\conda\conda\envs\tensorflow\Lib\site-packages\tensorflow\contrib\keras\python\keras\callbacks.py`

but the code I am running still does not recognise write_batch_performance as an argument.

**Comment by JulesGM** [comment]
you add this `TB` class to your code, instantiate one and pass it to the fit function under the callback argument when you call it, ie

```python
class TB(callbacks.TensorBoard):
    def __init__(self, log_every=1, **kwargs):
        super().__init__(**kwargs)
        self.log_every = log_every
        self.counter = 0
    
    def on_batch_end(self, batch, logs=None):
        self.counter+=1
        if self.counter%self.log_every==0:
            for name, value in logs.items():
                if name in ['batch', 'size']:
                    continue
                summary = tf.Summary()
                summary_value = summary.value.add()
                summary_value.simple_value = value.item()
                summary_value.tag = name
                self.writer.add_summary(summary, self.counter)
            self.writer.flush()
        
        super().on_batch_end(batch, logs)

# ... The rest of your code

batch_tensorboard_callback = TB()
model.fit(features, labels, validation=(features_va, labels_va), callbacks=batch_tensorboard_callback)
```

**Comment by JulesGM** [comment]
@RyanCodes44 don't change the code in callbacks, just instantiate the class in your code and use it as a callback like https://keras.io/callbacks/#example-recording-loss-history

**Comment by RyanCodes44** [comment]
Okay thanks guys!

**Comment by VladislavZavadskyy** [comment]
As another workaround, I've found helpful using
```python
tb_callback = TensorBoard()
tb_callback.on_batch_end = tb_callback.on_epoch_end
```
Much less code this way. One could also prevent calling code in `on_epoch_end` by setting it to lambda with `pass` inside.

**Comment by bersbersbers** [comment]
Regarding the code example by @sameermanek, @jpcenteno80, @VladislavZavadskyy (I have tried the latter): it seems to work fine, with two issues:
1. loss and metrics are written out for each batch, but val_loss and val_metrics are not (only once per epoch). Is this intended?
2. the on_epoch_end being still active (because I need it for val_loss etc, see 1), it writes out data where the step number is the running batch number, interspersed with data where the step number is the epoch number. This leads to artifacts: for example, with 2 batches per epoch, this is what I get:
![image](https://user-images.githubusercontent.com/12128514/39047795-83f3e146-449b-11e8-8722-db6e59558c66.png)
One can clearly see the trend of both per-epoch (left hull) and per-batch (right hull) curves, but it would be easier without the per-epoch data.

**Comment by bersbersbers** [comment]
3. Another issue revolves around the comparison of runs having different batch sizes - this will become difficult as the same number of epochs does not translate into the same number of steps. I believe in view of all these issues 2 and 3, it would make sense to write per-batch and per-epoch output to separate scalar streams (if that is the correct term) and not the same one.

**Comment by GuillaumeDesforges** [comment]
Bump, the PR seems idle. This feature would be really useful !

**Comment by DmitriiDenisov** [comment]
@jayanthc But in your case nothing will be written in val metrics

**Comment by DmitriiDenisov** [comment]
I actually suggest adding this function into ```class TB(callbacks.TensorBoard)```:
```    def on_epoch_end(self, epoch, logs=None):
        for name, value in logs.items():
            if (name in ['batch', 'size']) or ('val' not in name):
                continue
            summary = tf.Summary()
            summary_value = summary.value.add()
            summary_value.simple_value = value.item()
            summary_value.tag = name
            self.writer.add_summary(summary, epoch)
        self.writer.flush()```

**Comment by jayanthc** [comment]
@DmitriiDenisov: You are right. I ended up adding something similar to what you posted, to `on_epoch_end`.

**Comment by gabrieldemarmiesse** [comment]
This feature has been implemented and merged in master.

**Comment by ybagdasa** [comment]
When I set the update_freq parameter in tf.keras.callbacks.TensorBoard to 128 (update every 128 batches) it only affects the training loss/metrics. The validation loss/metrics are still plotted by epoch. This is my only enabled callback. 
![Screenshot from 2020-04-21 10-56-48](https://user-images.githubusercontent.com/5349813/79898168-a3e66900-83bf-11ea-8938-373271ba7a27.png)
I'm using tf version 2.2.0-dev20200331 and tb version 2.3.0a20200331 from a recently nightly release.

**Comment by GF-Huang** [comment]
So how to show the metrics based on batch? Not the epoch？
